src/core/bsearch.rkt

```diff
@@ -100,6 +101,20 @@
     (cons val (random-ref pts)))
   (apply mk-pcontext (cdr (batch-prepare-points evaluator new-sampler))))
 
+(define cache (make-hash))
+(define (cache-put! key value)
+        (hash-set! cache key value))
+(define (cache-get-prepend v expr macro)
+        (define key (cons expr v))
+        (define value (hash-ref cache key #f))
+        (if (eq? value #f) 
+            (begin 
+              (set! value (macro v))
+              (cache-put! key value))
+            (void))
+        value) 
```

Comment:
You can use `hash-ref!`, which is a helper function for exactly this pattern—get something from a hash table or compute it and save it for next time. Then you also won't need `cache-put!`